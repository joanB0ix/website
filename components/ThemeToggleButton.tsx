"use client";

import { MoonIcon, SunIcon } from "@heroicons/react/24/solid";
import { useEffect, useState } from "react";

export default function ThemeToggleButton() {
  const [isDarkMode, setIsDarkMode] = useState(false);

  useEffect(() => {
    const darkMode = window.matchMedia("(prefers-color-scheme: dark)").matches;
    setIsDarkMode(darkMode);
    document.documentElement.classList.toggle("dark", darkMode);
  }, []);

  const toggleTheme = () => {
    setIsDarkMode((prev) => !prev);
    document.documentElement.classList.toggle("dark");
  };

  return (
    <button
      onClick={toggleTheme}
      className="flex items-center justify-center h-10 w-10 rounded-full bg-gray-200 dark:bg-gray-800 transition"
      aria-label="Toggle Theme"
    >
      {isDarkMode ? (
        <SunIcon className="h-6 w-6 text-yellow-400" data-testid="sun-icon" />
      ) : (
        <MoonIcon className="h-6 w-6 text-blue-500" data-testid="moon-icon" />
      )}
    </button>
  );
}
