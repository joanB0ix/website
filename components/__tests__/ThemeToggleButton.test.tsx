import { fireEvent, render, screen } from "@testing-library/react";
import ThemeToggleButton from "../ThemeToggleButton";

beforeAll(() => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: query === "(prefers-color-scheme: light)",
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    })),
  });
});

describe("ThemeToggleButton", () => {
  it("Should render correctly and toggle theme", () => {
    render(<ThemeToggleButton />);

    const moonIcon = screen.getByTestId("moon-icon");
    expect(moonIcon).toBeInTheDocument();

    fireEvent.click(moonIcon);

    const sunIcon = screen.getByTestId("sun-icon");
    expect(sunIcon).toBeInTheDocument();
    expect(document.documentElement.classList.contains("dark")).toBe(true);

    fireEvent.click(sunIcon);

    expect(document.documentElement.classList.contains("dark")).toBe(false);
  });

  it("Should match snapshot", () => {
    const { asFragment } = render(<ThemeToggleButton />);
    expect(asFragment()).toMatchSnapshot();
  });
});
