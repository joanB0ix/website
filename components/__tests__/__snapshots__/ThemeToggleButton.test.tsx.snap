// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeToggleButton Should match snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="Toggle Theme"
    class="flex items-center justify-center h-10 w-10 rounded-full bg-gray-200 dark:bg-gray-800 transition"
  >
    <svg
      aria-hidden="true"
      class="h-6 w-6 text-blue-500"
      data-slot="icon"
      data-testid="moon-icon"
      fill="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M9.528 1.718a.75.75 0 0 1 .162.819A8.97 8.97 0 0 0 9 6a9 9 0 0 0 9 9 8.97 8.97 0 0 0 3.463-.69.75.75 0 0 1 .981.98 10.503 10.503 0 0 1-9.694 6.46c-5.799 0-10.5-4.7-10.5-10.5 0-4.368 2.667-8.112 6.46-9.694a.75.75 0 0 1 .818.162Z"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</DocumentFragment>
`;
